### PR TITLE
Add gradient for QR decomposition of wide matrices (nrows < ncols)

### PR DESCRIPTION
**Pull request by D-Roberts**
This PR adds the gradient calculation for the QR decomposition of real wide input matrices (num_rows < num_cols).  

The gradient for the QR decomposition was requested in issue [#6504](https://github.com/tensorflow/tensorflow/issues/6504) .
Currently Tf supports the QR gradient calculation for square and deep matrices (rows >= cols), when full_matrices = False.

There are TODO(pfau) items  to add the gradient calculation for wide matrices and for the cases with full_matrices=True in the test file [qr_op_test.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/qr_op_test.py) .

This PR resolves one and a half TODO items: 
- wide real input matrix (nrows < ncols) and full_matrices = False.
- wide real input matrix (nrows < ncols) and full_matrices = True.

Code is largely based on the following references:
- [Liao et al. Differentiable Programming Tensor Networks](https://journals.aps.org/prx/pdf/10.1103/PhysRevX.9.031041)
- [Ionescu et al. Matrix Backpropagation for Deep Networks with Structured Layers](https://www.cv-foundation.org/openaccess/content_iccv_2015/papers/Ionescu_Matrix_Backpropagation_for_ICCV_2015_paper.pdf)
- [Giles, M. An extended collection of matrix derivative results
for forward and reverse mode algorithmic
differentiation](https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf)

### CODE REVIEW

**Comment by D-Roberts** [comment]
Hi @rmlarsen can you please take a look and let me know your thoughts? thanks

**Review by ezhulenev — DISMISSED** [review]

**Comment by gbaned** [comment]
@D-Roberts Can you please check build failures. Thanks!

**Comment by D-Roberts** [comment]
@gbaned I pushed another commit that should hopefully fix the build failures.

**Review by ezhulenev — APPROVED** [review]

**Comment by D-Roberts** [comment]
@ezhulenev thank you for your review. @gbaned please let me know if there is anything else necessary to do here.
